src/providers/push/fcm/fcm.provider.ts

```diff
@@ -0,0 +1,128 @@
+        ) {
+          throw new HttpErrors.BadRequest(
+            'Message receiver, topic or condition not found in request !',
+          );
+        }
+
+        if (message.receiver.to.length > 500) {
+          throw new HttpErrors.BadRequest(
+            'Message receiver count cannot exceed 500 !',
+          );
+        }
+
+        /**
+         * This method is responsible to send all the required data to moile application
+         * The mobile device will recieve push notification.
+         * Push will be sent to the devices with registration token sent in receiver
+         * Notification object holds title, body and imageUrl
+         * FCM message must contain 2 attributes, i.e title and body
+         *
+         */
+
+        if (!message.subject) {
+          throw new HttpErrors.BadRequest('Message title not found !');
+        }
+
+        const promises: Promise<string | admin.messaging.BatchResponse>[] = [];
+
+        // These are the registration tokens for all devices which this message is intended for
+        const receiverTokens = message.receiver.to.map(item => item.id);
+
+        const standardNotifForFCM: admin.messaging.Notification = {
+          body: message.body,
+          title: message.subject,
+          imageUrl: message.options.imageUrl,
+        };
+
+        let msgToTransfer:
+          | admin.messaging.MulticastMessage
+          | TopicMessage
+          | ConditionMessage;
+
+        /**
+         * Message attributes for all kinds of messages
+         */
+        const generalMessageObj = {
+          notification: standardNotifForFCM,
+          android: message.options.android,
+          webpush: message.options.webpush,
+          apns: message.options.apns,
+          fcmOptions: message.options.fcmOptions,
+        };
+
+        /**
+         * if the receivers are empty check for topic or condition
+         * */
+        if (receiverTokens.length >= 1) {
+          msgToTransfer = {
+            tokens: receiverTokens,
+            ...generalMessageObj,
+          };
+          promises.push(
+            this.fcmService
+              .messaging()
+              .sendMulticast(msgToTransfer, (message.options.dryRun = false)),
+          );
+        } else if (message.options.topic) {
+          // Topic message
+          msgToTransfer = {
+            topic: message.options.topic,
+            ...generalMessageObj,
+          };
+          promises.push(
+            this.fcmService
+              .messaging()
+              .send(msgToTransfer, (message.options.dryRun = false)),
+          );
+        } else if (message.options.condition) {
+          // COndition message
+          msgToTransfer = {
+            condition: message.options.condition,
+            ...generalMessageObj,
+          };
+          promises.push(
+            this.fcmService
+              .messaging()
+              .send(msgToTransfer, (message.options.dryRun = false)),
```

Comment:
why this dryrun ?

Reply:
Whether or not the message should actually be sent. When set to `true`, allows developers to test a request without actually sending a message. When set to `false`, the message will be sent.